src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -30,14 +32,80 @@ public ArrayListProductDao() {
                 .findAny();
     }
 
+    private double countPercentage(int count, String description) {
+        return (double) count / description.split(" ").length;
+    }
+
+    private synchronized HashMap<Product, Integer> clearFrequency(HashMap<Product, Integer> frequency, int count) {
```

Comment:
HashMap approach may be a bit complicated. Do we really need the map here? We need the ordered list of products to be returned.

Could you please try something like:

```
ToIntFunction<Product> getNumberOfMatches = product -> (int) Arrays.stream(keywords)
                .filter(product.getDescription().toLowerCase()::contains)
                .count();
```

Here keywords are the words split by space.

We can then use this function to sort the product list:

```
return products.stream()
                .filter(...)
                .sorted(Comparator.comparingInt(getNumberOfMatches)...)
                .collect(...);
```